internals/overlord/logstate/syslog/syslog.go

```diff
@@ -0,0 +1,298 @@
+package syslog
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/url"
+	"os"
+	"sort"
+	"time"
+
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const (
+	defaultMaxRequestEntries    = 100
+	defaultSyslogInitialBackoff = 100 * time.Millisecond
+	defaultSyslogMaxBackoff     = 10 * time.Second
+	defaultDialTimeout          = 10 * time.Second
+	canonicalPrivEnterpriseNum  = 28978
+)
+
+// Syslog Priority values - see RFC 5424 6.2.1
+const (
+	FacilityUserLevelMessage = 1
+	SeverityInformational    = 6
+)
+
+type ClientOptions struct {
+	MaxRequestEntries    int
+	TargetName           string
+	Location             string
+	SyslogInitialBackoff time.Duration
+	SyslogMaxBackoff     time.Duration
+	DialTimeout          time.Duration
+}
+
+type entryWithService struct {
+	service   string
+	Priority  int
+	Version   int
+	Timestamp string
+	PID       string
+	MsgID     string
+	Message   string
+}
+
+type Client struct {
+	options *ClientOptions
+
+	// To store log entries, keep a buffer of size 2*MaxRequestEntries with a
+	// sliding window "entries" of size MaxRequestEntries.
+	buffer  []entryWithService
+	entries []entryWithService
+
+	// Store the custom labels(syslog's structured-data) for each service
+	labels map[string]string
+
+	// connection info
+	conn          net.Conn
+	address       *url.URL
+	waitReconnect time.Duration
+	closed        bool
+	hostname      string
+	data          bytes.Buffer
```

Comment:
These two aren't "connection info", so should be separated from those with a blank line.

Also, the name "data" is a bit too generic (everything is data), so perhaps change to `sendBuf`.